File: apps/mobile/targets/positions-widget/Sources/Utilities/RangeMath.swift
import Foundation

enum RangeMath {
  /// Band width clamps. Floor keeps the band visible against the thumb on
  /// micro-ranges; ceiling leaves bar margin for out-of-range overshoot.
  static let minBandWidthPct: Double = 0.20
  static let maxBandWidthPct: Double = 0.70

  /// Sigmoid in `log10(tickWidth)` space — center anchored at `log10(1000)`
  /// (≈ ±5% price band, a typical concentrated LP range). Wider position →
  /// wider band, monotonically, no hard plateau.
  static let bandWidthLogCenter: Double = 3.0
  static let bandWidthLogSpread: Double = 1.0

  /// Overshoot compression. As `|current - bound| / rangeWidth` grows the
  /// thumb travels through the side margin, asymptotically against the edge.
  static let overshootScale: Double = 1.5

  struct BarPositions {
    let liquidityLeftPct: Double
    let liquidityWidthPct: Double
    let thumbPct: Double
    let inRange: Bool
  }

  static func barPositions(currentTick: Int, tickLower: Int, tickUpper: Int) -> BarPositions {
    let rangeWidth = max(1, Double(tickUpper - tickLower))
    let bandWidth = bandWidthFor(tickWidth: rangeWidth)
    let bandLeftPct = (1 - bandWidth) / 2
    let bandRightPct = bandLeftPct + bandWidth

    let currentPos = Double(currentTick - tickLower) / rangeWidth
    let inRange = currentPos >= 0 && currentPos <= 1

    let thumbPct: Double
    if inRange {
      thumbPct = bandLeftPct + currentPos * bandWidth
    } else if currentPos < 0 {
      let overshoot = -currentPos
      let traveled = bandLeftPct * (1 - exp(-overshoot / overshootScale))
      thumbPct = bandLeftPct - traveled
    } else {
      let overshoot = currentPos - 1
      let traveled = (1 - bandRightPct) * (1 - exp(-overshoot / overshootScale))
      thumbPct = bandRightPct + traveled
    }

    return BarPositions(
      liquidityLeftPct: bandLeftPct,
      liquidityWidthPct: bandWidth,
      thumbPct: thumbPct,
      inRange: inRange
    )
  }

  private static func bandWidthFor(tickWidth: Double) -> Double {
    let logTicks = log10(max(1, tickWidth))
    let normalized = (logTicks - bandWidthLogCenter) / bandWidthLogSpread
    let sigmoid = 1 / (1 + exp(-normalized))
    return minBandWidthPct + sigmoid * (maxBandWidthPct - minBandWidthPct)
  }
}

File: apps/mobile/modules/widget-bridge/android/src/main/java/expo/modules/widgetbridge/widget/util/RangeMath.kt
package expo.modules.widgetbridge.widget.util

import kotlin.math.exp
import kotlin.math.log10
import kotlin.math.max

object RangeMath {
  private const val MIN_BAND_WIDTH_PCT = 0.20
  private const val MAX_BAND_WIDTH_PCT = 0.70
  private const val BAND_WIDTH_LOG_CENTER = 3.0
  private const val BAND_WIDTH_LOG_SPREAD = 1.0
  private const val OVERSHOOT_SCALE = 1.5

  data class BarPositions(
    val liquidityLeftPct: Double,
    val liquidityWidthPct: Double,
    val thumbPct: Double,
    val inRange: Boolean,
  )

  fun barPositions(currentTick: Int, tickLower: Int, tickUpper: Int): BarPositions {
    val rangeWidth = max(1.0, (tickUpper - tickLower).toDouble())
    val bandWidth = bandWidthFor(rangeWidth)
    val bandLeftPct = (1 - bandWidth) / 2
    val bandRightPct = bandLeftPct + bandWidth

    val currentPos = (currentTick - tickLower).toDouble() / rangeWidth
    val inRange = currentPos in 0.0..1.0

    val thumbPct = when {
      inRange -> bandLeftPct + currentPos * bandWidth
      currentPos < 0 -> {
        val overshoot = -currentPos
        val traveled = bandLeftPct * (1 - exp(-overshoot / OVERSHOOT_SCALE))
        bandLeftPct - traveled
      }
      else -> {
        val overshoot = currentPos - 1
        val traveled = (1 - bandRightPct) * (1 - exp(-overshoot / OVERSHOOT_SCALE))
        bandRightPct + traveled
      }
    }

    return BarPositions(bandLeftPct, bandWidth, thumbPct, inRange)
  }

  private fun bandWidthFor(tickWidth: Double): Double {
    val logTicks = log10(max(1.0, tickWidth))
    val normalized = (logTicks - BAND_WIDTH_LOG_CENTER) / BAND_WIDTH_LOG_SPREAD
    val sigmoid = 1 / (1 + exp(-normalized))
    return MIN_BAND_WIDTH_PCT + sigmoid * (MAX_BAND_WIDTH_PCT - MIN_BAND_WIDTH_PCT)
  }
}
